doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,49 @@
+# Design Doc: Execute the Program with Multi Thread
+
+In PaddlePaddle, the user could declare **wich** operators will be
+execute and **how much** threads will be run in Python code.
+
+## Python Interface
+
+We can specify the thread count with `parallel.do(thread_count=N)`,
+the following PaddlePaddle program shows the usage of the Parallel operator:
+
+```python
+import paddle.v2.fluid as fluid
+N=4
+parallel = fluid.parallel()
+
+x = minibatch([10, 20, 30, 40, 50])
+y = var("y")
+label = var(1)
+w = var("w")
+b = var("b")
+
+fluid.split(x, N)
+
+with parallel.do(thread_count=N):
+    y_predict = fluid.fc(input=x, size=1)
+    cost = fluid.layers.square_error_cost(input=y_predict, label=y)
+    avg_cost = fluid.layers.mean(x=cost)
```

Comment:
I understand you have put a lot of time thinking about the API, but as a user looking at the API I have several questions:

- `x` is split by `fluid.split(x, N)`, how about `y_predict`, `y` and `avg_cost`, do the user need to split it?
- If the `avg_cost` is split into `N`, can the user set `thread_count` to `M` in:
    ```
    with parallel.do(thread_count=M):
      sgd_optimizer.minimize(avg_cost)
    ```
- why should the user call `fluid.merge_grad_avg(x, N)`, what will happen if the user forget to call it, and how can the user debug this problem?

From the above questions, I think the Python API for `parallel.Do` is very hard to design and hard to understand for the user. I briefly talked with @wangkuiyi last week, and we agreed that the Python API is lower priority comparing to the auto transpiler.

So maybe we should postpone the Python API design for `parallel.Do` for now (maybe focus more on transpiler in this PR?), and focus on the automatic program transpiler. In this way the user don't even need to worry about the Python API for `parallel.Do`. If there are the power user who want to use the Python API (I highly doubt it), we can add the Python API until then.

Reply:
> why should the user call fluid.merge_grad_avg(x, N), what will happen if the user forget to call it, and how can the user debug this problem?

If we allow the user to describe a **Program **, maybe user need to describe some details like merge gradient from the result with different threads.

> So maybe we should postpone the Python API design for parallel.Do for now (maybe focus more on transpiler in this PR?),

I got it, maybe you're right, use `parallel.Do` to describe the Program is not easy enough for the user.

Reply:
Thanks!